File: CTCI/Arrays and Strings/isUnique.java
import java.util.BitSet;
import java.util.Arrays;

public class isUnique{
	
	//Array for ascii value approach. Time: O(n) Space: 0(n)
	static boolean isUnique1(String str){
		BitSet alphabet = new BitSet(128);
		for(int i = 0; i < str.length(); i++){
			if(alphabet.get(str.charAt(i)) == true)
				return false;
			alphabet.flip(str.charAt(i));
		}
		return true;
	}

	//Sort String then compare neighboring letters.
	static boolean isUnique2(String str){
		char[] sorted = str.toCharArray();
		Arrays.sort(sorted);
		for(int i = 1; i < sorted.length; i++){
			if(sorted[i] == sorted[i-1])
				return false;
		}
		return true;
	}

	public static void main(String[] args){
		System.out.println(isUnique2("Helo"));
		return;
	}
}

File: CTCI/Arrays and Strings/isUnique.py
import time

#Create list size of alphabet (ascii values) assumes  with each index representing a letter and initial values at 0. Increment per letter found. If at any point an index is greater than 1 return False. Else return True.

def isUnique1(String):
    
    alphabet = [0] * 128
    
    for c in String:    
        alphabet[ord(c)] +=1
        
        if alphabet[ord(c)] > 1:
            return False
    
    return True

#Similar to above, but uses built in dictionary. Time O(n) Space O(n)
def isUnique2(String):
    
    dictionary = {}
    for c in String:
        #How long does this operation take
        if c in dictonary:
            return False
        else:
            dictionary.update({c:1})
    return True

#Sorted and compare each index to prev. Time: O(nlogn) Space: O(0)
def isUnique3(String):
    #Not sure what happens if xrange(1,1) or xrange(1,0). Will test with time tests.
    if len(String) == 1 or len(String) == 0:
        return True

    sortedString = sorted(String)
    for i in xrange(1,sortedString):
        if sortedString[i-1] == sortedString[i]:
            return False
    return True

#No sort and no extra space. Compare each ele to each other. Time: O(n^2) Space: O(0)
def isUnique4(String):
    for c in String:
        for c2 in String:
            if c == c2:
                return False
    return True

start = time.clock()
print(isUnique1("Helo"))
end = time.clock()
print(end-start)
